startup/95-sample-custom.py

```diff
@@ -0,0 +1,3260 @@
+        incident_angles: [0.05, 0.10.2];
+        exposure_time: in second, corresponding to the detector setting;
+        tiling: 'x', 'y', 'xy', None
+        
+        '''
+        if detector == 'BOTH':
+            detector = ['SAXS', 'WAXS']
+            if type(exposure_time) == int:
+                exposure_time = [exposure_time, exposure_time]
+        self.measure_setting['detector'] = detector
+        self.measure_setting['incident_angles'] = incident_angles
+        self.measure_setting['exposure_time'] = exposure_time
+        self.measure_setting['tiling'] = tiling
+        self.detector = detector
+        self.incident_angles = incident_angles
+        self.exposure_time = exposure_time
+        self.tiling = tiling
+        
+        return self.measure_setting
+    
+    def alignSamples(self, range=None, step=0, align_step=0, x_offset=0, verbosity=3, **md):
+        '''Iterates through the samples on the holder, aligning each one.'''
+        
+        if step<=0:
+            get_beamline().modeAlignment()
+        
+        
+        if step<=5:
+            for sample in self.getSamples(range=range):
+                sample.gotoOrigin(['x','y','th'])
+                sample.gotoOrigin(['x'])
+                sample.xr(x_offset)
+                sample.align(step=align_step, reflection_angle=0.12)
+            
+            
+        if step<=10:
+            if verbosity>=3:
+                print('Alignment complete.')
+                for i, sample in enumerate(self.getSamples()):
+                    print('Sample {} ({})'.format(i+1, sample.name))
+                    print(sample.save_state())
+
+
+    def alignSamplesQuick(self, range=None, step=0, x_offset=0, verbosity=3, **md):
+        '''Iterates through the samples on the holder, aligning each one.'''
+        
+        if step<=0:
+            get_beamline().modeAlignment()
+        
+        
+        if step<=5:
+            for sample in self.getSamples(range=range):
+                sample.gotoOrigin(['x','y','th'])
+                sample.gotoOrigin(['x'])
+                sample.xr(x_offset)
+                sample.alignQuick(reflection_angle=0.12) #=0.07)
+            
+            
+        if step<=10:
+            if verbosity>=3:
+                print('Alignment complete.')
+                for i, sample in enumerate(self.getSamples()):
+                    print('Sample {} ({})'.format(i+1, sample.name))
+                    print(sample.save_state())
+                
+
+    def alignSamplesVeryQuick(self, range=None, step=0, x_offset=0, verbosity=3, **md):
+        '''Iterates through the samples on the holder, aligning each one.'''
+        
+        if step<=0:
+            get_beamline().modeAlignment()
+            beam.on()
+            caput('XF:11BMB-ES{Det:SAXS}:cam1:AcquireTime', 0.25)
+            caput('XF:11BMB-ES{Det:SAXS}:cam1:AcquirePeriod', 0.30)
+        
+        
+        if step<=5:
```

Comment:
I think this might be incorrect. This code will repeat 5 times, and the code where step=10 will repeat 10 times.

Reply:
This pattern seems to be used in many places. Is it doing what they expect?

Reply:
Ruipeng says that they are using the "step" mechanism to leave these functions and only execute code they want to run - for example, in this block, if step==0, it will run the whole function, step==5, only run the first 2.

I have suggested we go over some of these mechanisms that they have developed and see if we can standardize things so that their code will be shorter and easier to follow.